axoupdater/src/test/helpers.rs

```diff
@@ -0,0 +1,216 @@
+        .replace("VERSION", version)
+        .replace("RELEASE_TYPE", &release_type.to_string())
+}
+
+/// Generates an install receipt given the specified fields and writes it to disk at the appropriate location in `config_path`. The path to the new file is returned.
+#[allow(clippy::too_many_arguments)]
+pub fn write_receipt(
+    app_name: &str,
+    package: &str,
+    owner: &str,
+    binaries: &[String],
+    version: &str,
+    prefix: &Path,
+    config_path: &PathBuf,
+    release_type: &ReleaseSourceType,
+) -> std::io::Result<PathBuf> {
+    let contents = install_receipt(
+        app_name,
+        package,
+        owner,
+        binaries,
+        version,
+        &prefix.to_string_lossy(),
+        release_type,
+    );
+    let receipt_name = config_path.join(format!("{package}-receipt.json"));
+    std::fs::create_dir_all(config_path)?;
+    std::fs::write(&receipt_name, contents)?;
+
+    Ok(receipt_name)
+}
+
+/// The arguments used for `perform_runtest`.
+pub struct RuntestArgs {
+    /// The name of the app being tested.
+    pub app_name: String,
+    /// The name of the package/workspace being tested. In GitHub terms, this is the "name" of the owner/name repo format.
+    pub package: String,
+    /// The owner of the package being tested. In GitHub terms, this is the "owner" of the owner/name repo format.
+    pub owner: String,
+    /// The path to the executable being tested, usually the one from the `CARGO_BIN_EXE_<name>` environment variable.
+    pub bin: PathBuf,
+    /// A list of all binaries installed by the app being tested.
+    pub binaries: Vec<String>,
+    /// The default path to which the app is installed. For most apps, this is $HOME/.cargo.
+    pub installation_path: PathBuf,
+    /// The arguments taken by the binary being tested.
+    ///
+    /// For example, for cargo dist, it's called as `cargo dist selfupdate --skip-init`, so this value is `vec!["dist", "selfupdate", "--skip-init"]`.
+    pub args: Vec<String>,
+    /// The type of release to test, either GitHub or Axo Releases.
+    pub release_type: ReleaseSourceType,
+}
+
+/// Actually installs your app and runs its updater.
+/// For detailed information on the arguments, see [`RuntestArgs`][].
+///
+/// This function performs several assertions of its own, then returns the path to which the binary was expected to have been installed in order to allow the caller to perform additional tests or assertions.
+/// Because it writes to real files outside a temporary directory, it's highly recommended that this only becalled within CI builds.
+pub fn perform_runtest(runtest_args: &RuntestArgs) -> PathBuf {
+    let RuntestArgs {
+        app_name,
+        package,
+        owner,
+        bin,
+        binaries,
+        installation_path,
+        args,
+        release_type,
+    } = runtest_args;
+
+    let basename = bin.file_name().unwrap();
+
+    let app_home = &installation_path.join("bin");
```

Comment:
Hmm this doesn't seem correct.

Reply:
This is me trying to force us back into a more FHS-ish structure - the prefix is the root, the bindir is a subdirectory of the prefix. So in this context `installation_path` is `~/.cargo`, the bin goes in `~/.cargo/bin`. Over in cargo-dist, it sets

```rust
let home = homedir::get_my_home().unwrap().unwrap();
installation_path = home.join(".cargo")
```

Reply:
Right but if they set anything but `~/.cargo/` here, slapping `bin` on the end is just wrong (I think this probably wants to be an enum or something, so the user doesn't need to know how this works... but maybe we can just not make it configurable for now to avoid thinking about it)

Reply:
Made it unconfigurable for the moment, and we can do something cleverer in the future as it comes up. Also added a comment noting it's `~/.cargo/bin`.